nexus/src/app/external_endpoints.rs

```diff
@@ -0,0 +1,1052 @@
+use nexus_db_queries::db::datastore::Discoverability;
+use nexus_db_queries::db::DataStore;
+use nexus_types::identity::Resource;
+use omicron_common::api::external::http_pagination::PaginatedBy;
+use omicron_common::api::external::DataPageParams;
+use omicron_common::api::external::Error;
+use omicron_common::api::external::Name as ExternalName;
+use omicron_common::bail_unless;
+use openssl::pkey::PKey;
+use openssl::x509::X509;
+use rustls::sign::CertifiedKey;
+use serde::Serialize;
+use serde_with::SerializeDisplay;
+use std::collections::btree_map::Entry;
+use std::collections::BTreeMap;
+use std::fmt;
+use std::num::NonZeroU32;
+use std::sync::Arc;
+use thiserror::Error;
+use tokio::sync::watch;
+use uuid::Uuid;
+
+/// Describes the set of external endpoints, organized by DNS name
+///
+/// This data structure provides a quick way to determine which Silo and TLS
+/// certificate(s) make sense for an incoming request, based on the TLS
+/// session's SNI (DNS name).  See module-level docs for details.
+///
+/// This object provides no interfaces outside this module.  It's only used by
+/// the `NexusCertResolver` that's also in this module.
+///
+/// This structure impls `Serialize` only so that background tasks can easily
+/// present the latest configuration that they've found (e.g., via a debug API)
+#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
+pub struct ExternalEndpoints {
+    by_dns_name: BTreeMap<String, Arc<ExternalEndpoint>>,
+    warnings: Vec<ExternalEndpointError>,
+}
+
+impl ExternalEndpoints {
+    /// Assemble a list of Silos, TLS certificates, and external DNS zones into
+    /// a structure that we can use for quickly figuring out which Silo and TLS
+    /// certificates are associated with each incoming DNS name
+    fn new(
+        silos: Vec<nexus_db_model::Silo>,
+        certs: Vec<Certificate>,
+        external_dns_zones: Vec<nexus_db_model::DnsZone>,
+    ) -> ExternalEndpoints {
+        // We want to avoid failing this operation even if we encounter problems
+        // because we want to serve as many DNS certificates as we can (so that
+        // an operator has a chance of fixing any problems that do exist).
+        // Instead of returning any errors, keep track of any issues as
+        // warnings.
+        let mut warnings = vec![];
+
+        // Compute a mapping from external DNS name to Silo id.  Detect any
+        // duplicates and leave them out (but report them).  There should not
+        // be any duplicates since the DNS names are constructed from the
+        // (unique) Silo names.  Even if we support aliases in the future, they
+        // will presumably need to be unique, too.
+        let silo_names: BTreeMap<Uuid, ExternalName> = silos
+            .iter()
+            .map(|db_silo| (db_silo.id(), db_silo.name().clone()))
+            .collect();
+        let mut dns_names: BTreeMap<String, Uuid> = BTreeMap::new();
+        for z in external_dns_zones {
```

Comment:
Well...good question.  There is no way that this would happen today.  When I originally did this work, I tried to make things work with multiple external DNS zones, only as a way to be able to rename the external DNS zone gracefully.  So I've tried to avoid having code assume that there's only one.  (I'm not sure this is the right tradeoff, but it's generally been easy enough and it's already what we do elsewhere.)